sidecarhttp/http_api.go

```diff
@@ -289,6 +290,46 @@ func (s *SidecarApi) stateHandler(response http.ResponseWriter, req *http.Reques
 	}
 }
 
+// drainServiceHandler instructs Sidecar to set the status of a given service
+// instance to DRAINING. This allows us to decomission the given service
+// instance and let it sit around for a short amount of time, so it can finish
+// processing the requests that are still in flight.
+func (s *SidecarApi) drainServiceHandler(response http.ResponseWriter, req *http.Request, params map[string]string) {
+	defer req.Body.Close()
+
+	if req.Method != http.MethodPost {
+		sendJsonError(response, 400, fmt.Sprintf("Bad request - Method %q not allowed", req.Method))
+		return
+	}
+
+	serviceID, ok := params["id"]
+	if !ok {
+		sendJsonError(response, 404, "Not Found - No service ID provided")
+		return
+	}
+
+	if s.state == nil {
+		sendJsonError(response, 500, "Internal Server Error - Something went terribly wrong")
+		return
+	}
+
+	svc, err := s.state.GetServiceByID(serviceID)
+	if err != nil {
+		sendJsonError(response, 404, fmt.Sprintf("Not Found - Service ID %q not found", serviceID))
+		return
+	}
+
+	svc.Updated = time.Now()
+	svc.Status = service.DRAINING
+	s.state.UpdateService(svc)
```

Comment:
I think you probably should call `AddServiceEntry()` here instead so that it doesn't end up getting queued behind any other updates.